### PR TITLE
gh-106246: Allow the use of unions as match patterns

### PR DESCRIPTION
**Pull request by tmke8**
This didn't require many code changes. Though I'm not entirely sure about code style and whether I added enough tests.

* Issue: gh-118524

* Issue: gh-106246

### CODE REVIEW

File: Lib/test/test_patma.py [hunk 1]
```diff
@@ -3361,6 +3370,31 @@ class A:
                     w = 0
         self.assertIsNone(w)
 
+    def test_union_type_postional_subpattern(self):
```
Comment (sylvainmouquet) [inline Lib/test/test_patma.py:3373]: english is not my mother tongue, but `postional` must not to be replaced by `positional`?

File: Lib/test/test_patma.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import dataclasses
 import enum
 import inspect
+from re import I
```
Comment (JelleZijlstra) [inline Lib/test/test_patma.py:6]: accidental?

File: Lib/test/test_patma.py [hunk 3]
```diff
@@ -2886,6 +2887,14 @@ class B(A): ...
                         h = 1
                 self.assertEqual(h, 1)
 
+    def test_patma_union_type(self):
+        IntOrStr = int | str
+        x = 0
+        match x:
+            case IntOrStr():
+                x = 1
```
Comment (JelleZijlstra) [inline Lib/test/test_patma.py:2895]: Add a test that involves a union that does *not* match.

**Comment by nineteendo** [comment]
Please rename the title to use gh-106246.

File: Python/ceval.c [hunk 4]
```diff
@@ -470,6 +471,16 @@ _PyEval_MatchClass(PyThreadState *tstate, PyObject *subject, PyObject *type,
     if (PyObject_IsInstance(subject, type) <= 0) {
         return NULL;
     }
+    // Subpatterns are not supported for union types:
+    if (_PyUnion_Check(type)) {
```
Comment (sobolevn) [inline Python/ceval.c:475]: I don't like the fact that we call `_PyUnion_Check(type)` twice for no real reason.

Comment (nineteendo) [reply Python/ceval.c:475]: Could we save the result in a variable? Or would it cleaner to handle unions first?

**Comment by randolf-scholz** [comment]
@sobolevn @JelleZijlstra @sylvainmouquet @tmke8 I pushed a variant of this pull request that addresses the raised issues and also support subpatterns #118644.

**Pull request closed without merging**